Sender will retry forever while MetadataUtils.getOneAvailableTabletServerNode() throw FlussRuntimeException

### Search before asking

- [x] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.


### Fluss version

main (development)

### Please describe the bug 🐞

Sender will retry forever while `MetadataUtils.getOneAvailableTabletServerNode()` throw `FlussRuntimeException`, it cannot be recovery:

![Image](https://github.com/user-attachments/assets/fe114501-4256-4a5b-9f87-f1cfe025b91d)

### Solution

Although now I don't know why it failed here, but when the exception `Caused by: com.alibaba.fluss.exception.FlussRuntimeException: no alive tablet server in cluster` is thrown, we should be able to recover by using the `tryToInitializeCluster` method.

Update:
After checking the root cause, it turns out that when one `tabletServer` is killed, there is a period during which the `metadataUpdater` request metadata from the server that does not include the killed `tabletServer`. As a result, `metadataUpdater.getTabletServer()` will return null. However, we currently do not have protection in place for this scenario, leading to a `NullPointerException (NPE) `when creating the `Gateway`. The affected behaviors include `Sender`, `LogFetcher#sendFetchRequest()`, and `LookupSender#sendLookups()`. Therefore, it is necessary to systematically reconsider the design of the client metadata as mentioned in issue: #483

![Image](https://github.com/user-attachments/assets/0667548a-219e-43b3-934b-0f3615ce9f7f)

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/metadata/MetadataUpdater.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/metadata/MetadataUpdater.java
@@ -245,7 +245,8 @@ public class MetadataUpdater {
         updateMetadata(updateTablePaths, updatePartitionPath, null);
     }
 
-    private void updateMetadata(
+    @VisibleForTesting
+    protected void updateMetadata(
             @Nullable Set<TablePath> tablePaths,
             @Nullable Collection<PhysicalTablePath> tablePartitionNames,
             @Nullable Collection<Long> tablePartitionIds) {

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/table/scanner/log/LogFetcher.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/table/scanner/log/LogFetcher.java
@@ -27,6 +27,7 @@ import com.alibaba.fluss.cluster.ServerNode;
 import com.alibaba.fluss.config.ConfigOptions;
 import com.alibaba.fluss.config.Configuration;
 import com.alibaba.fluss.exception.InvalidMetadataException;
+import com.alibaba.fluss.exception.LeaderNotAvailableException;
 import com.alibaba.fluss.fs.FsPath;
 import com.alibaba.fluss.metadata.PhysicalTablePath;
 import com.alibaba.fluss.metadata.TableBucket;
@@ -198,16 +199,11 @@ public class LogFetcher implements Closeable {
         // TODO cache the tablet server gateway.
         ServerNode destinationNode = metadataUpdater.getTabletServer(destination);
         if (destinationNode == null) {
-            // invalid the metadata to request metadata update
-            LOG.warn(
-                    "Server {} is not found in metadata cache, "
-                            + "going to request metadata update.",
-                    destination);
-            invalidTableOrPartitions(tableOrPartitionsInFetchRequest);
-            LOG.debug(
-                    "Removing pending request for node: {} due to the destination node is not found in metadata cache",
-                    destination);
-            nodesWithPendingFetchRequests.remove(destination);
+            handleFetchLogException(
+                    destination,
+                    tableOrPartitionsInFetchRequest,
+                    new LeaderNotAvailableException(
+                            "Server " + destination + " is not found in metadata cache."));
         } else {
             TabletServerGateway gateway =
                     GatewayClientProxy.createGatewayProxy(
@@ -218,13 +214,15 @@ public class LogFetcher implements Closeable {
 
             gateway.fetchLog(fetchLogRequest)
                     .whenComplete(
-                            (fetchLogResponse, e) ->
+                            (fetchLogResponse, e) -> {
+                                if (e != null) {
+                                    handleFetchLogException(
+                                            destination, tableOrPartitionsInFetchRequest, e);
+                                } else {
                                     handleFetchLogResponse(
-                                            destination,
-                                            requestStartTime,
-                                            fetchLogResponse,
-                                            tableOrPartitionsInFetchRequest,
-                                            e));
+                                            destination, requestStartTime, fetchLogResponse);
+                                }
+                            });
         }
     }
 
@@ -270,30 +268,30 @@ public class LogFetcher implements Closeable {
         metadataUpdater.invalidPhysicalTableBucketMeta(physicalTablePaths);
     }
 
+    private void handleFetchLogException(
+            int destination, TableOrPartitions tableOrPartitionsInFetchRequest, Throwable e) {
+        try {
+            LOG.error("Failed to fetch log from node {}", destination, e);
+            // if is invalid metadata exception, we need to clear table bucket meta
+            // to enable another round of log fetch to request new medata
+            if (e instanceof InvalidMetadataException) {
+                LOG.warn(
+                        "Invalid metadata error in fetch log request. "
+                                + "Going to request metadata update.",
+                        e);
+                invalidTableOrPartitions(tableOrPartitionsInFetchRequest);
+            }
+        } catch (Exception ex) {
+            LOG.debug("Removing pending request for node: {}", destination);
+            nodesWithPendingFetchRequests.remove(destination);
+        }
+    }
+
     /** Implements the core logic for a successful fetch log response. */
     private synchronized void handleFetchLogResponse(
-            int destination,
-            long requestStartTime,
-            FetchLogResponse fetchLogResponse,
-            TableOrPartitions tableOrPartitionsInFetchRequest,
-            @Nullable Throwable e) {
+            int destination, long requestStartTime, FetchLogResponse fetchLogResponse) {
         try {
             if (isClosed) {
-                return;
-            }
-
-            if (e != null) {
-                LOG.error("Failed to fetch log from node {}", destination, e);
-
-                // if is invalid metadata exception, we need to clear table bucket meta
-                // to enable another round of log fetch to request new medata
-                if (e instanceof InvalidMetadataException) {
-                    LOG.warn(
-                            "Received invalid metadata error in fetch log request. "
-                                    + "Going to request metadata update.",
-                            e);
-                    invalidTableOrPartitions(tableOrPartitionsInFetchRequest);
-                }
                 return;
             }
 

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/table/scanner/log/LogFetcher.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/table/scanner/log/LogFetcher.java
@@ -281,7 +281,7 @@ public class LogFetcher implements Closeable {
                         e);
                 invalidTableOrPartitions(tableOrPartitionsInFetchRequest);
             }
-        } catch (Exception ex) {
+        } finally {
             LOG.debug("Removing pending request for node: {}", destination);
             nodesWithPendingFetchRequests.remove(destination);
         }

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/table/scanner/log/LogFetcher.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/table/scanner/log/LogFetcher.java
@@ -23,6 +23,7 @@ import com.alibaba.fluss.client.metrics.ScannerMetricGroup;
 import com.alibaba.fluss.client.table.scanner.RemoteFileDownloader;
 import com.alibaba.fluss.client.table.scanner.ScanRecord;
 import com.alibaba.fluss.cluster.BucketLocation;
+import com.alibaba.fluss.cluster.ServerNode;
 import com.alibaba.fluss.config.ConfigOptions;
 import com.alibaba.fluss.config.Configuration;
 import com.alibaba.fluss.exception.InvalidMetadataException;
@@ -192,28 +193,39 @@ public class LogFetcher implements Closeable {
     }
 
     private void sendFetchRequest(int destination, FetchLogRequest fetchLogRequest) {
-        // TODO cache the tablet server gateway.
-        TabletServerGateway gateway =
-                GatewayClientProxy.createGatewayProxy(
-                        () -> metadataUpdater.getTabletServer(destination),
-                        rpcClient,
-                        TabletServerGateway.class);
-
-        final long requestStartTime = System.currentTimeMillis();
-        scannerMetricGroup.fetchRequestCount().inc();
-
         TableOrPartitions tableOrPartitionsInFetchRequest =
                 getTableOrPartitionsInFetchRequest(fetchLogRequest);
+        // TODO cache the tablet server gateway.
+        ServerNode destinationNode = metadataUpdater.getTabletServer(destination);
+        if (destinationNode == null) {
+            // invalid the metadata to request metadata update
+            LOG.warn(
+                    "Server {} is not found in metadata cache, "
+                            + "going to request metadata update.",
+                    destination);
+            invalidTableOrPartitions(tableOrPartitionsInFetchRequest);
+            LOG.debug(
+                    "Removing pending request for node: {} due to the destination node is not found in metadata cache",
+                    destination);
+            nodesWithPendingFetchRequests.remove(destination);
+        } else {
+            TabletServerGateway gateway =
+                    GatewayClientProxy.createGatewayProxy(
+                            () -> destinationNode, rpcClient, TabletServerGateway.class);
 
-        gateway.fetchLog(fetchLogRequest)
-                .whenComplete(
-                        (fetchLogResponse, e) ->
-                                handleFetchLogResponse(
-                                        destination,
-                                        requestStartTime,
-                                        fetchLogResponse,
-                                        tableOrPartitionsInFetchRequest,
-                                        e));
+            final long requestStartTime = System.currentTimeMillis();
+            scannerMetricGroup.fetchRequestCount().inc();
+
+            gateway.fetchLog(fetchLogRequest)
+                    .whenComplete(
+                            (fetchLogResponse, e) ->
+                                    handleFetchLogResponse(
+                                            destination,
+                                            requestStartTime,
+                                            fetchLogResponse,
+                                            tableOrPartitionsInFetchRequest,
+                                            e));
+        }
     }
 
     private TableOrPartitions getTableOrPartitionsInFetchRequest(FetchLogRequest fetchLogRequest) {
@@ -251,6 +263,13 @@ public class LogFetcher implements Closeable {
         }
     }
 
+    private void invalidTableOrPartitions(TableOrPartitions tableOrPartitions) {
+        Set<PhysicalTablePath> physicalTablePaths =
+                metadataUpdater.getPhysicalTablePathByIds(
+                        tableOrPartitions.tableIds, tableOrPartitions.tablePartitions);
+        metadataUpdater.invalidPhysicalTableBucketMeta(physicalTablePaths);
+    }
+
     /** Implements the core logic for a successful fetch log response. */
     private synchronized void handleFetchLogResponse(
             int destination,
@@ -269,15 +288,11 @@ public class LogFetcher implements Closeable {
                 // if is invalid metadata exception, we need to clear table bucket meta
                 // to enable another round of log fetch to request new medata
                 if (e instanceof InvalidMetadataException) {
-                    Set<PhysicalTablePath> physicalTablePaths =
-                            metadataUpdater.getPhysicalTablePathByIds(
-                                    tableOrPartitionsInFetchRequest.tableIds,
-                                    tableOrPartitionsInFetchRequest.tablePartitions);
                     LOG.warn(
                             "Received invalid metadata error in fetch log request. "
                                     + "Going to request metadata update.",
                             e);
-                    metadataUpdater.invalidPhysicalTableBucketMeta(physicalTablePaths);
+                    invalidTableOrPartitions(tableOrPartitionsInFetchRequest);
                 }
                 return;
             }

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/utils/MetadataUtils.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/utils/MetadataUtils.java
@@ -271,6 +271,7 @@ public class MetadataUtils {
         return aliveTabletServers.get(offset);
     }
 
+    @Nullable
     private static ServerNode getCoordinatorServer(MetadataResponse response) {
         if (!response.hasCoordinatorServer()) {
             return null;

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/utils/MetadataUtils.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/utils/MetadataUtils.java
@@ -21,6 +21,7 @@ import com.alibaba.fluss.cluster.Cluster;
 import com.alibaba.fluss.cluster.ServerNode;
 import com.alibaba.fluss.cluster.ServerType;
 import com.alibaba.fluss.exception.FlussRuntimeException;
+import com.alibaba.fluss.exception.StaleMetadataException;
 import com.alibaba.fluss.metadata.PhysicalTablePath;
 import com.alibaba.fluss.metadata.TableBucket;
 import com.alibaba.fluss.metadata.TableDescriptor;
@@ -103,11 +104,17 @@ public class MetadataUtils {
         return gateway.metadata(metadataRequest)
                 .thenApply(
                         response -> {
-                            ServerNode coordinatorServer = getCoordinatorServer(response);
-
                             // Update the alive table servers.
                             Map<Integer, ServerNode> newAliveTabletServers =
                                     getAliveTabletServers(response);
+                            // when talking to the startup tablet
+                            // server, it maybe receive empty metadata, we'll consider it as
+                            // stale metadata and throw StaleMetadataException which will cause
+                            // to retry later.
+                            if (newAliveTabletServers.isEmpty()) {
+                                throw new StaleMetadataException("Alive tablet server is empty.");
+                            }
+                            ServerNode coordinatorServer = getCoordinatorServer(response);
 
                             Map<TablePath, Long> newTablePathToTableId;
                             Map<TablePath, TableInfo> newTablePathToTableInfo;
@@ -266,7 +273,7 @@ public class MetadataUtils {
 
     private static ServerNode getCoordinatorServer(MetadataResponse response) {
         if (!response.hasCoordinatorServer()) {
-            throw new FlussRuntimeException("coordinator server is not found");
+            return null;
         } else {
             PbServerNode protoServerNode = response.getCoordinatorServer();
             return new ServerNode(

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/write/Sender.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/write/Sender.java
@@ -22,6 +22,7 @@ import com.alibaba.fluss.client.metrics.WriterMetricGroup;
 import com.alibaba.fluss.client.write.RecordAccumulator.ReadyCheckResult;
 import com.alibaba.fluss.cluster.ServerNode;
 import com.alibaba.fluss.exception.InvalidMetadataException;
+import com.alibaba.fluss.exception.LeaderNotAvailableException;
 import com.alibaba.fluss.exception.OutOfOrderSequenceException;
 import com.alibaba.fluss.exception.RetriableException;
 import com.alibaba.fluss.exception.UnknownTableOrBucketException;
@@ -331,14 +332,10 @@ public class Sender implements Runnable {
 
         ServerNode destinationNode = metadataUpdater.getTabletServer(destination);
         if (destinationNode == null) {
-            LOG.warn(
-                    "Server {} is not found in metadata cache, "
-                            + "going to request metadata update.",
-                    destination);
-            // invalidate the metadata to request metadata
-            invalidBucketMetadata(batches);
-            // requeue the write batch
-            batches.forEach(this::reEnqueueBatch);
+            handleWriteRequestException(
+                    new LeaderNotAvailableException(
+                            "Server " + destination + " is not found in metadata cache."),
+                    recordsByBucket);
         } else {
             TabletServerGateway gateway = metadataUpdater.newTabletServerClientForNode(destination);
             writeBatchByTable.forEach(
@@ -448,14 +445,6 @@ public class Sender implements Runnable {
             } else {
                 completeBatch(writeBatch);
             }
-        }
-        metadataUpdater.invalidPhysicalTableBucketMeta(invalidMetadataTablesSet);
-    }
-
-    private void invalidBucketMetadata(List<WriteBatch> writeBatches) {
-        Set<PhysicalTablePath> invalidMetadataTablesSet = new HashSet<>();
-        for (WriteBatch writeBatch : writeBatches) {
-            invalidMetadataTablesSet.add(writeBatch.physicalTablePath());
         }
         metadataUpdater.invalidPhysicalTableBucketMeta(invalidMetadataTablesSet);
     }

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/write/Sender.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/write/Sender.java
@@ -329,25 +329,38 @@ public class Sender implements Runnable {
                             .add(batch);
                 });
 
-        TabletServerGateway gateway = metadataUpdater.newTabletServerClientForNode(destination);
-        writeBatchByTable.forEach(
-                (tableId, writeBatches) -> {
-                    TableInfo tableInfo = metadataUpdater.getTableInfoOrElseThrow(tableId);
-                    if (tableInfo.hasPrimaryKey()) {
-                        sendPutKvRequestAndHandleResponse(
-                                gateway,
-                                makePutKvRequest(tableId, acks, maxRequestTimeoutMs, writeBatches),
-                                tableId,
-                                recordsByBucket);
-                    } else {
-                        sendProduceLogRequestAndHandleResponse(
-                                gateway,
-                                makeProduceLogRequest(
-                                        tableId, acks, maxRequestTimeoutMs, writeBatches),
-                                tableId,
-                                recordsByBucket);
-                    }
-                });
+        ServerNode destinationNode = metadataUpdater.getTabletServer(destination);
+        if (destinationNode == null) {
+            LOG.warn(
+                    "Server {} is not found in metadata cache, "
+                            + "going to request metadata update.",
+                    destination);
+            // invalidate the metadata to request metadata
+            invalidBucketMetadata(batches);
+            // requeue the write batch
+            batches.forEach(this::reEnqueueBatch);
+        } else {
+            TabletServerGateway gateway = metadataUpdater.newTabletServerClientForNode(destination);
+            writeBatchByTable.forEach(
+                    (tableId, writeBatches) -> {
+                        TableInfo tableInfo = metadataUpdater.getTableInfoOrElseThrow(tableId);
+                        if (tableInfo.hasPrimaryKey()) {
+                            sendPutKvRequestAndHandleResponse(
+                                    gateway,
+                                    makePutKvRequest(
+                                            tableId, acks, maxRequestTimeoutMs, writeBatches),
+                                    tableId,
+                                    recordsByBucket);
+                        } else {
+                            sendProduceLogRequestAndHandleResponse(
+                                    gateway,
+                                    makeProduceLogRequest(
+                                            tableId, acks, maxRequestTimeoutMs, writeBatches),
+                                    tableId,
+                                    recordsByBucket);
+                        }
+                    });
+        }
     }
 
     private void sendProduceLogRequestAndHandleResponse(
@@ -435,6 +448,14 @@ public class Sender implements Runnable {
             } else {
                 completeBatch(writeBatch);
             }
+        }
+        metadataUpdater.invalidPhysicalTableBucketMeta(invalidMetadataTablesSet);
+    }
+
+    private void invalidBucketMetadata(List<WriteBatch> writeBatches) {
+        Set<PhysicalTablePath> invalidMetadataTablesSet = new HashSet<>();
+        for (WriteBatch writeBatch : writeBatches) {
+            invalidMetadataTablesSet.add(writeBatch.physicalTablePath());
         }
         metadataUpdater.invalidPhysicalTableBucketMeta(invalidMetadataTablesSet);
     }

--- a/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
+++ b/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
@@ -34,7 +34,7 @@ class MetadataUpdaterTest {
 
     @RegisterExtension
     public static final FlussClusterExtension FLUSS_CLUSTER_EXTENSION =
-            FlussClusterExtension.builder().setNumOfTabletServers(1).build();
+            FlussClusterExtension.builder().setNumOfTabletServers(2).build();
 
     @Test
     void testUpdateWithEmptyMetadataResponse() throws Exception {
@@ -47,6 +47,7 @@ class MetadataUpdaterTest {
         Cluster cluster = metadataUpdater.getCluster();
 
         List<ServerNode> expectedServerNodes = FLUSS_CLUSTER_EXTENSION.getTabletServerNodes();
+        assertThat(expectedServerNodes).hasSize(2);
         assertThat(cluster.getAliveTabletServerList()).isEqualTo(expectedServerNodes);
 
         // then, stop coordinator server, can still update metadata
@@ -56,14 +57,16 @@ class MetadataUpdaterTest {
 
         // start a new tablet server, the tablet server will return empty metadata
         // response since no coordinator server to send newest metadata to the tablet server
-        FLUSS_CLUSTER_EXTENSION.startTabletServer(1);
+        int newServerId = 2;
+        FLUSS_CLUSTER_EXTENSION.startTabletServer(newServerId);
 
         // we mock a new cluster with only server 1 so that it'll only send request
         // to server 1, which will return empty resonate
         Cluster newCluster =
                 new Cluster(
                         Collections.singletonMap(
-                                1, FLUSS_CLUSTER_EXTENSION.getTabletServerNodes().get(1)),
+                                newServerId,
+                                FLUSS_CLUSTER_EXTENSION.getTabletServerNodes().get(newServerId)),
                         null,
                         Collections.emptyMap(),
                         Collections.emptyMap(),
@@ -74,6 +77,7 @@ class MetadataUpdaterTest {
         // shouldn't update metadata to empty since the empty metadata will be ignored
         metadataUpdater.updateMetadata(null, null, null);
         assertThat(metadataUpdater.getCluster().getAliveTabletServers())
-                .isEqualTo(newCluster.getAliveTabletServers());
+                .isEqualTo(newCluster.getAliveTabletServers())
+                .hasSize(1);
     }
 }

--- a/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
+++ b/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
@@ -1,0 +1,79 @@
+/*
+ * Copyright (c) 2025 Alibaba Group Holding Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.fluss.client.metadata;
+
+import com.alibaba.fluss.cluster.Cluster;
+import com.alibaba.fluss.cluster.ServerNode;
+import com.alibaba.fluss.rpc.RpcClient;
+import com.alibaba.fluss.server.testutils.FlussClusterExtension;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for update metadata of {@link MetadataUpdater}. */
+class MetadataUpdaterTest {
+
+    @RegisterExtension
+    public static final FlussClusterExtension FLUSS_CLUSTER_EXTENSION =
+            FlussClusterExtension.builder().setNumOfTabletServers(1).build();
+
+    @Test
+    void testUpdateWithEmptyMetadataResponse() throws Exception {
+        RpcClient rpcClient = FLUSS_CLUSTER_EXTENSION.getRpcClient();
+        MetadataUpdater metadataUpdater =
+                new MetadataUpdater(FLUSS_CLUSTER_EXTENSION.getClientConfig(), rpcClient);
+
+        // update metadata
+        metadataUpdater.updateMetadata(null, null, null);
+        Cluster cluster = metadataUpdater.getCluster();
+
+        List<ServerNode> expectedServerNodes = FLUSS_CLUSTER_EXTENSION.getTabletServerNodes();
+        assertThat(cluster.getAliveTabletServerList()).isEqualTo(expectedServerNodes);
+
+        // then, stop coordinator server, can still update metadata
+        FLUSS_CLUSTER_EXTENSION.stopCoordinatorServer();
+        metadataUpdater.updateMetadata(null, null, null);
+        assertThat(cluster.getAliveTabletServerList()).isEqualTo(expectedServerNodes);
+
+        // start a new tablet server, the tablet server will return empty metadata
+        // response since no coordinator server to send newest metadata to the tablet server
+        FLUSS_CLUSTER_EXTENSION.startTabletServer(1);
+
+        // we mock a new cluster with only server 1 so that it'll only send request
+        // to server 1, which will return empty resonate
+        Cluster newCluster =
+                new Cluster(
+                        Collections.singletonMap(
+                                1, FLUSS_CLUSTER_EXTENSION.getTabletServerNodes().get(1)),
+                        null,
+                        Collections.emptyMap(),
+                        Collections.emptyMap(),
+                        Collections.emptyMap(),
+                        Collections.emptyMap());
+
+        metadataUpdater = new MetadataUpdater(rpcClient, newCluster);
+        // shouldn't update metadata to empty since the empty metadata will be ignored
+        metadataUpdater.updateMetadata(null, null, null);
+        assertThat(metadataUpdater.getCluster().getAliveTabletServers())
+                .isEqualTo(newCluster.getAliveTabletServers());
+    }
+}

--- a/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/TestingMetadataUpdater.java
+++ b/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/TestingMetadataUpdater.java
@@ -72,6 +72,10 @@ public class TestingMetadataUpdater extends MetadataUpdater {
         }
     }
 
+    public void updateCluster(Cluster cluster) {
+        this.cluster = cluster;
+    }
+
     @Override
     public void checkAndUpdateTableMetadata(Set<TablePath> tablePaths) {
         Set<TablePath> needUpdateTablePaths =

--- a/fluss-client/src/test/java/com/alibaba/fluss/client/table/scanner/log/LogFetcherTest.java
+++ b/fluss-client/src/test/java/com/alibaba/fluss/client/table/scanner/log/LogFetcherTest.java
@@ -21,6 +21,8 @@ import com.alibaba.fluss.client.metadata.MetadataUpdater;
 import com.alibaba.fluss.client.metrics.TestingScannerMetricGroup;
 import com.alibaba.fluss.client.table.scanner.RemoteFileDownloader;
 import com.alibaba.fluss.client.table.scanner.ScanRecord;
+import com.alibaba.fluss.cluster.Cluster;
+import com.alibaba.fluss.cluster.ServerNode;
 import com.alibaba.fluss.metadata.TableBucket;
 import com.alibaba.fluss.record.MemoryLogRecords;
 import com.alibaba.fluss.rpc.RpcClient;
@@ -119,6 +121,67 @@ public class LogFetcherTest extends ClientToServerITCaseBase {
         // after collect fetch, the fetcher is empty.
         assertThat(logFetcher.hasAvailableFetches()).isFalse();
         assertThat(logFetcher.getCompletedFetchesSize()).isEqualTo(0);
+    }
+
+    @Test
+    void testFetchWhenDestinationIsNullInMetadata() throws Exception {
+        TableBucket tb0 = new TableBucket(tableId, bucketId0);
+        addRecordsToBucket(tb0, genMemoryLogRecordsByObject(DATA1), 0L);
+
+        RpcClient rpcClient = FLUSS_CLUSTER_EXTENSION.getRpcClient();
+        MetadataUpdater metadataUpdater = new MetadataUpdater(clientConf, rpcClient);
+        metadataUpdater.checkAndUpdateTableMetadata(Collections.singleton(DATA1_TABLE_PATH));
+
+        int leaderNode = metadataUpdater.leaderFor(tb0);
+
+        // now, remove leader nodd ,so that fetch destination
+        // server node is null
+        Cluster oldCluster = metadataUpdater.getCluster();
+        Map<Integer, ServerNode> aliveTabletServersById =
+                new HashMap<>(oldCluster.getAliveTabletServers());
+        aliveTabletServersById.remove(leaderNode);
+        Cluster newCluster =
+                new Cluster(
+                        aliveTabletServersById,
+                        oldCluster.getCoordinatorServer(),
+                        oldCluster.getBucketLocationsByPath(),
+                        oldCluster.getTableIdByPath(),
+                        oldCluster.getPartitionIdByPath(),
+                        oldCluster.getTableInfoByPath());
+        metadataUpdater = new MetadataUpdater(rpcClient, newCluster);
+
+        LogScannerStatus logScannerStatus = new LogScannerStatus();
+        logScannerStatus.assignScanBuckets(Collections.singletonMap(tb0, 0L));
+
+        LogFetcher logFetcher =
+                new LogFetcher(
+                        DATA1_TABLE_INFO,
+                        null,
+                        rpcClient,
+                        logScannerStatus,
+                        clientConf,
+                        metadataUpdater,
+                        TestingScannerMetricGroup.newInstance(),
+                        new RemoteFileDownloader(1));
+
+        // send fetches to fetch data, should have no available fetch.
+        logFetcher.sendFetches();
+        assertThat(logFetcher.hasAvailableFetches()).isFalse();
+
+        // then fetches again, should have available fetch.
+        // first send fetch is for update metadata
+        logFetcher.sendFetches();
+        // second send fetch will do real fetch data
+        logFetcher.sendFetches();
+        retry(
+                Duration.ofMinutes(1),
+                () -> {
+                    assertThat(logFetcher.hasAvailableFetches()).isTrue();
+                    assertThat(logFetcher.getCompletedFetchesSize()).isEqualTo(1);
+                });
+        Map<TableBucket, List<ScanRecord>> records = logFetcher.collectFetch();
+        assertThat(records.size()).isEqualTo(1);
+        assertThat(records.get(tb0).size()).isEqualTo(10);
     }
 
     private void addRecordsToBucket(

--- a/fluss-client/src/test/java/com/alibaba/fluss/client/write/SenderTest.java
+++ b/fluss-client/src/test/java/com/alibaba/fluss/client/write/SenderTest.java
@@ -18,6 +18,7 @@ package com.alibaba.fluss.client.write;
 
 import com.alibaba.fluss.client.metadata.TestingMetadataUpdater;
 import com.alibaba.fluss.client.metrics.TestingWriterMetricGroup;
+import com.alibaba.fluss.cluster.Cluster;
 import com.alibaba.fluss.cluster.ServerNode;
 import com.alibaba.fluss.config.ConfigOptions;
 import com.alibaba.fluss.config.Configuration;
@@ -42,6 +43,7 @@ import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Deque;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -594,6 +596,53 @@ final class SenderTest {
         assertThat(idempotenceManager.lastAckedBatchSequence(tb1)).isEqualTo(Optional.of(0));
         assertThat(future1.isDone()).isTrue();
         assertThat(future1.get()).isNull();
+    }
+
+    @Test
+    void testSendWhenDestinationIsNullInMetadata() throws Exception {
+        long offset = 0;
+        CompletableFuture<Exception> future = new CompletableFuture<>();
+        appendToAccumulator(tb1, row(1, "a"), future::complete);
+
+        int leaderNode = metadataUpdater.leaderFor(tb1);
+        // now, remove leader node ,so that send destination
+        // server node is null
+        Cluster oldCluster = metadataUpdater.getCluster();
+        Map<Integer, ServerNode> aliveTabletServersById =
+                new HashMap<>(oldCluster.getAliveTabletServers());
+        aliveTabletServersById.remove(leaderNode);
+        Cluster newCluster =
+                new Cluster(
+                        aliveTabletServersById,
+                        oldCluster.getCoordinatorServer(),
+                        oldCluster.getBucketLocationsByPath(),
+                        oldCluster.getTableIdByPath(),
+                        oldCluster.getPartitionIdByPath(),
+                        oldCluster.getTableInfoByPath());
+
+        metadataUpdater.updateCluster(newCluster);
+
+        sender.runOnce();
+        // should be no inflight batches
+        assertThat(sender.numOfInFlightBatches(tb1)).isEqualTo(0);
+
+        // the bucket location should be empty for the bucket since we'll invalid it
+        // when send to a null destination
+        assertThat(metadataUpdater.getCluster().getBucketLocation(tb1)).isEmpty();
+
+        // update with old cluster to mock a metadata update
+        metadataUpdater.updateCluster(oldCluster);
+
+        // send again, should send successfully
+        sender.runOnce();
+        assertThat(sender.numOfInFlightBatches(tb1)).isEqualTo(1);
+
+        finishProduceLogRequest(tb1, 0, createProduceLogResponse(tb1, offset, 1));
+
+        // send again, should send nothing since no batch in queue
+        sender.runOnce();
+        assertThat(sender.numOfInFlightBatches(tb1)).isEqualTo(0);
+        assertThat(future.get()).isNull();
     }
 
     private TestingMetadataUpdater initializeMetadataUpdater() {

--- a/fluss-common/src/main/java/com/alibaba/fluss/exception/LeaderNotAvailableException.java
+++ b/fluss-common/src/main/java/com/alibaba/fluss/exception/LeaderNotAvailableException.java
@@ -1,0 +1,39 @@
+/*
+ *  Copyright (c) 2025 Alibaba Group Holding Ltd.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.alibaba.fluss.exception;
+
+import com.alibaba.fluss.annotation.PublicEvolving;
+
+/**
+ * There is no currently available leader for the given partition (either because a leadership
+ * election is in progress or because all replicas are down).
+ *
+ * @since 0.7
+ */
+@PublicEvolving
+public class LeaderNotAvailableException extends InvalidMetadataException {
+
+    private static final long serialVersionUID = 1L;
+
+    public LeaderNotAvailableException(String message) {
+        super(message);
+    }
+
+    public LeaderNotAvailableException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}

--- a/fluss-common/src/main/java/com/alibaba/fluss/exception/StaleMetadataException.java
+++ b/fluss-common/src/main/java/com/alibaba/fluss/exception/StaleMetadataException.java
@@ -16,10 +16,15 @@
 
 package com.alibaba.fluss.exception;
 
+import com.alibaba.fluss.annotation.PublicEvolving;
+
 /**
  * Thrown when current metadata cannot be used. This is often used to trigger a new update metadata
  * request.
+ *
+ * @since 0.7
  */
+@PublicEvolving
 public class StaleMetadataException extends InvalidMetadataException {
 
     private static final long serialVersionUID = 1L;

--- a/fluss-common/src/main/java/com/alibaba/fluss/exception/StaleMetadataException.java
+++ b/fluss-common/src/main/java/com/alibaba/fluss/exception/StaleMetadataException.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright (c) 2025 Alibaba Group Holding Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.fluss.exception;
+
+/**
+ * Thrown when current metadata cannot be used. This is often used to trigger a new update metadata
+ * request.
+ */
+public class StaleMetadataException extends InvalidMetadataException {
+
+    private static final long serialVersionUID = 1L;
+
+    public StaleMetadataException(String message) {
+        super(message);
+    }
+}

--- a/fluss-rpc/src/main/java/com/alibaba/fluss/rpc/protocol/Errors.java
+++ b/fluss-rpc/src/main/java/com/alibaba/fluss/rpc/protocol/Errors.java
@@ -38,6 +38,7 @@ import com.alibaba.fluss.exception.InvalidUpdateVersionException;
 import com.alibaba.fluss.exception.KvSnapshotNotExistException;
 import com.alibaba.fluss.exception.KvStorageException;
 import com.alibaba.fluss.exception.LakeStorageNotConfiguredException;
+import com.alibaba.fluss.exception.LeaderNotAvailableException;
 import com.alibaba.fluss.exception.LogOffsetOutOfRangeException;
 import com.alibaba.fluss.exception.LogStorageException;
 import com.alibaba.fluss.exception.NetworkException;
@@ -182,7 +183,11 @@ public enum Errors {
     PARTITION_ALREADY_EXISTS(
             42, "The partition already exists.", PartitionAlreadyExistsException::new),
     PARTITION_SPEC_INVALID_EXCEPTION(
-            43, "The partition spec is invalid.", InvalidPartitionException::new);
+            43, "The partition spec is invalid.", InvalidPartitionException::new),
+    LEADER_NOT_AVAILABLE_EXCEPTION(
+            44,
+            "There is no currently available leader for the given partition.",
+            LeaderNotAvailableException::new);
 
     private static final Logger LOG = LoggerFactory.getLogger(Errors.class);
 

